FirebasePerformance/Sources/FPRNanoPbUtils.m

```diff
@@ -0,0 +1,502 @@
+  networkMetric.response_payload_bytes = trace.responseSize;
+  networkMetric.has_response_payload_bytes = true;
+
+  networkMetric.http_response_code = trace.responseCode;
+  networkMetric.has_http_response_code = true;
+  networkMetric.response_content_type = FPREncodeString(trace.responseContentType);
+
+  if (trace.responseError) {
+    networkMetric.network_client_error_reason =
+        firebase_perf_v1_NetworkRequestMetric_NetworkClientErrorReason_GENERIC_CLIENT_ERROR;
+    networkMetric.has_network_client_error_reason = true;
+  }
+
+  NSTimeInterval requestTimeUs =
+      USEC_PER_SEC *
+      [trace timeIntervalBetweenCheckpointState:FPRNetworkTraceCheckpointStateInitiated
+                                       andState:FPRNetworkTraceCheckpointStateRequestCompleted];
+  if (requestTimeUs > 0) {
+    networkMetric.time_to_request_completed_us = requestTimeUs;
+    networkMetric.has_time_to_request_completed_us = true;
+  }
+
+  NSTimeInterval responseIntiationTimeUs =
+      USEC_PER_SEC *
+      [trace timeIntervalBetweenCheckpointState:FPRNetworkTraceCheckpointStateInitiated
+                                       andState:FPRNetworkTraceCheckpointStateResponseReceived];
+  if (responseIntiationTimeUs > 0) {
+    networkMetric.time_to_response_initiated_us = responseIntiationTimeUs;
+    networkMetric.has_time_to_response_initiated_us = true;
+  }
+
+  NSTimeInterval responseCompletedUs =
+      USEC_PER_SEC *
+      [trace timeIntervalBetweenCheckpointState:FPRNetworkTraceCheckpointStateInitiated
+                                       andState:FPRNetworkTraceCheckpointStateResponseCompleted];
+  if (responseCompletedUs > 0) {
+    networkMetric.time_to_response_completed_us = responseCompletedUs;
+    networkMetric.has_time_to_response_completed_us = true;
+  }
+
+  // Filling custom attributes
+  NSDictionary<NSString *, NSString *> *attributes = [trace.attributes mutableCopy];
+  networkMetric.custom_attributes_count = attributes.count;
+  networkMetric.custom_attributes =
+      (struct _firebase_perf_v1_NetworkRequestMetric_CustomAttributesEntry *)
+          FPREncodeStringToStringMap(attributes);
+
+  // Filling session details
+  NSArray<FPRSessionDetails *> *orderedSessions = FPRMakeFirstSessionVerbose(trace.sessions);
+  networkMetric.perf_sessions_count = [orderedSessions count];
+  networkMetric.perf_sessions =
+      FPREncodePerfSessions(orderedSessions, networkMetric.perf_sessions_count);
+
+  return networkMetric;
+}
+
+firebase_perf_v1_GaugeMetric GetGaugeMetric(NSArray *gaugeData, NSString *sessionId) {
+  firebase_perf_v1_GaugeMetric gaugeMetric = firebase_perf_v1_GaugeMetric_init_default;
+  gaugeMetric.session_id = FPREncodeString(sessionId);
+
+  __block NSInteger cpuReadingsCount = 0;
+  __block NSInteger memoryReadingsCount = 0;
+
+  struct _firebase_perf_v1_CpuMetricReading *cpuReadings =
+      calloc([gaugeData count], sizeof(firebase_perf_v1_CpuMetricReading));
```

Comment:
If we are anyways going to reallocate at the end of counting the readings, should we start with less memory request first and later expand it? Instead of [gaugeData count], can we use just 1?

Reply:
Good point! Expanding one by one could be expensive though according https://stackoverflow.com/questions/6727136/performance-impact-of-realloc

Reply:
Sounds good.